src/QsCompiler/Transformations/CallGraph/CallGraphBase.cs

```diff
@@ -0,0 +1,255 @@
+        protected CallGraphEdgeBase(QsQualifiedName fromCallableName, QsQualifiedName toCallableName, DataTypes.Range referenceRange)
+        {
+            if (fromCallableName is null)
+            {
+                throw new ArgumentNullException(nameof(fromCallableName));
+            }
+
+            if (toCallableName is null)
+            {
+                throw new ArgumentNullException(nameof(toCallableName));
+            }
+
+            if (referenceRange is null)
+            {
+                throw new ArgumentNullException(nameof(referenceRange));
+            }
+
+            this.FromCallableName = fromCallableName;
+            this.ToCallableName = toCallableName;
+            this.ReferenceRange = referenceRange;
+        }
+
+        /// <summary>
+        /// Determines if the object is the same as the given edge, ignoring the
+        /// ordering of key-value pairs in the type parameter dictionaries.
+        /// </summary>
+        public bool Equals(CallGraphEdgeBase edge) =>
+            this.FromCallableName.Equals(edge.FromCallableName)
+            && this.ToCallableName.Equals(edge.ToCallableName)
+            && this.ReferenceRange.Equals(edge.ReferenceRange);
+    }
+
+    /// <summary>
+    /// Base class for call graph node types.
+    /// </summary>
+    public abstract class CallGraphNodeBase : IEquatable<CallGraphNodeBase>
+    {
+        /// <summary>
+        /// The name of the represented callable.
+        /// </summary>
+        public QsQualifiedName CallableName { get; }
+
+        /// <summary>
+        /// Base constructor for call graph nodes. Initializes CallableName.
+        /// Throws an ArgumentNullException if argument is null.
+        /// </summary>
+        protected CallGraphNodeBase(QsQualifiedName callableName)
+        {
+            if (callableName is null)
+            {
+                throw new ArgumentException(nameof(callableName));
+            }
+
+            this.CallableName = callableName;
+        }
+
+        /// <inheritdoc/>
+        public override bool Equals(object obj)
+        {
+            return obj is CallGraphNodeBase && this.Equals((CallGraphNodeBase)obj);
+        }
+
+        /// <inheritdoc/>
+        public bool Equals(CallGraphNodeBase other) =>
+            this.CallableName.Equals(other.CallableName);
+
+        /// <inheritdoc/>
+        public override int GetHashCode()
+        {
+            HashCode hash = default;
+            hash.Add(this.CallableName);
+            return hash.ToHashCode();
+        }
+    }
+
+    /// <summary>
+    /// Base class for call graph types.
+    /// </summary>
+    public abstract class CallGraphBase<TNode, TEdge>
+        where TNode : CallGraphNodeBase
+        where TEdge : CallGraphEdgeBase
+    {
+        // Static Elements
+
+        /// <summary>
+        /// Returns an empty dependency for a node.
+        /// </summary>
+        protected static ILookup<TNode, TEdge> EmptyDependency() =>
```

Comment:
This doesn't seem to be used outside the class:

```suggestion
        private static ILookup<TNode, TEdge> EmptyDependency() =>
```